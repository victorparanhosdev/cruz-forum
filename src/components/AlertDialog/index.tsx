'use client'
import {
  DialogActionTrigger,
  DialogBody,
  DialogCloseTrigger,
  DialogContent,
  DialogFooter,
  DialogHeader,
  DialogRoot,
  DialogTitle,
  DialogTrigger,
} from '@/components/ui/dialog'
import { ComponentProps, ReactNode, useState } from 'react'
import { Button } from '@/components'
import { toaster } from '@/components/ui/toaster'
import { CircleNotch } from '@phosphor-icons/react'
import { useRouter } from 'next/navigation'

interface DeleteTopicResponse {
  message?: string
  error?: string
}

interface AlertDialogProps extends ComponentProps<typeof DialogTrigger> {
  children: ReactNode
  topicSlug: number
  onDeleteTopic: (topicSlug: number) => Promise<DeleteTopicResponse>
}

export const AlertDialog = ({
  topicSlug,
  onDeleteTopic,
  children,
  ...props
}: AlertDialogProps) => {
  const [isOpen, setIsOpen] = useState(false)
  const [isLoading, setIsLoading] = useState(false)
  const router = useRouter()

  async function handleDeleteTopic() {
    setIsLoading(true)

    await onDeleteTopic(topicSlug)
      .then((res) => {
        if (res.error) {
          setIsLoading(false)
          setIsOpen(false)

          return toaster.error({
            title: 'Error',
            description: res.error,
            duration: 3000,
            type: 'error',
          })
        }

        router.refresh()
        setIsLoading(false)
        setIsOpen(false)

        toaster.success({
          title: 'Deletado com sucesso!',
          description: res?.message || 'O tópico foi removido com sucesso.',
          duration: 3000,
        })
      })
      .catch(console.error)
  }

  const handleDialogClose = ({ open }: { open: boolean }) => {
    setIsOpen(open)
  }

  return (
    <DialogRoot
      open={isOpen}
      onOpenChange={handleDialogClose}
      placement="center"
      role="alertdialog"
    >
      <DialogTrigger {...props} className="flex cursor-pointer items-center">
        {children}
      </DialogTrigger>
      <DialogContent className="rounded-lg bg-gray-950" padding={6}>
        <DialogHeader padding={0} marginBottom={4}>
          <DialogTitle className="text-2xl font-medium">
            Excluir Tópico
          </DialogTitle>
        </DialogHeader>

        <DialogBody padding={0}>
          <h1 className="text-base text-zinc-300">
            Tem certeza de que deseja excluir o tópico nome do tópico? Essa ação
            é irreversível e não poderá ser desfeita.
          </h1>
        </DialogBody>

        <DialogFooter padding={0} marginTop={6}>
          <DialogActionTrigger asChild>
            <Button state="outline-negative">Cancelar</Button>
          </DialogActionTrigger>

          {isLoading ? (
            <Button className="min-h-11 min-w-[101.16px]">
              <CircleNotch
                className="h-full w-full animate-spin text-white"
                size={20}
              />
            </Button>
          ) : (
            <Button type="button" onClick={handleDeleteTopic}>
              Continue
            </Button>
          )}
        </DialogFooter>

        <DialogCloseTrigger />
      </DialogContent>
    </DialogRoot>
  )
}
