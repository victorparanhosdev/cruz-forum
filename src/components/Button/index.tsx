import React, { ComponentType, forwardRef } from 'react'
import { twMerge } from 'tailwind-merge'
import { tv, type VariantProps } from 'tailwind-variants'
import { type IconProps } from '@phosphor-icons/react'

const buttonStyles = tv({
  base: 'flex items-center justify-center gap-2 rounded-lg text-base font-medium transition',
  variants: {
    state: {
      default:
        'bg-green-950 px-4 py-2.5 enabled:hover:bg-green-800 disabled:opacity-70',
      'outline-negative':
        'border border-red-900 bg-none px-4 py-2 enabled:hover:bg-error-100 disabled:opacity-70 disabled:bg-error-200',
      menu: 'w-full justify-start bg-none px-4 py-2 text-start hover:bg-hover-btn-menu_card data-[state=true]:bg-green-900 data-[state=true]:font-bold',
      transparent: 'gap-1 py-2 hover:enabled:text-green-200',
    },
  },
  defaultVariants: {
    state: 'default',
  },
})

interface ButtonProps
  extends VariantProps<typeof buttonStyles>,
    React.ButtonHTMLAttributes<HTMLButtonElement> {
  isActive?: boolean
  children: React.ReactNode
  iconLeft?: ComponentType<IconProps> | undefined
  iconRight?: ComponentType<IconProps> | undefined
  classNameIcon?: string
}

export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  (
    {
      state,
      isActive = false,
      children,
      className,
      disabled,
      iconLeft: IconLeft,
      iconRight: IconRight,
      classNameIcon,
      ...props
    },
    ref,
  ) => {
    const styles = buttonStyles({ state })
    return (
      <button
        ref={ref}
        data-state={isActive ? 'true' : undefined}
        aria-pressed={isActive ? 'true' : undefined}
        aria-disabled={disabled ? 'true' : undefined}
        disabled={disabled ? true : undefined}
        className={twMerge(styles, className)}
        {...props}
      >
        {IconLeft ? (
          <IconLeft
            data-testid="icon-left"
            aria-hidden="true"
            data-weight={isActive ? 'bold' : 'regular'}
            size={24}
            weight={isActive ? 'bold' : 'regular'}
            className={classNameIcon}
          />
        ) : undefined}
        <span className="whitespace-nowrap text-base">{children}</span>
        {IconRight ? (
          <IconRight
            data-testid="icon-right"
            aria-hidden="true"
            data-weight={isActive ? 'bold' : 'regular'}
            size={24}
            weight={isActive ? 'bold' : 'regular'}
            className={classNameIcon}
          />
        ) : undefined}
      </button>
    )
  },
)

Button.displayName = 'Button'
