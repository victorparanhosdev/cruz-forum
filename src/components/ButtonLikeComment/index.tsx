'use client'
import { Heart } from '@phosphor-icons/react/dist/ssr'
import { ButtonHTMLAttributes, useState } from 'react'
import { toaster } from '../ui/toaster'
import { motion } from 'framer-motion'

interface LikeButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  likes: number
  commentId: string
  isLike: boolean
}

interface HeartIconProps {
  isActive: boolean
  isHovered: boolean
}

const HeartIcon = ({ isActive, isHovered }: HeartIconProps) => {
  return (
    <motion.div
      initial={false}
      animate={{ scale: isActive || isHovered ? 1.1 : 1 }}
      transition={{ type: 'spring', stiffness: 300, damping: 20 }}
    >
      <Heart
        className={`${isActive ? 'text-green-600' : isHovered ? 'text-green-600' : 'text-white'}`}
        size={20}
        aria-hidden="true"
        weight={isActive || isHovered ? 'fill' : 'regular'}
      />
    </motion.div>
  )
}

export const ButtonLikeComment = ({
  likes,
  isLike,
  commentId,
  ...props
}: LikeButtonProps) => {
  const [isLiked, setIsLiked] = useState(isLike)
  const [likeCount, setLikeCount] = useState<number>(likes)
  const [isHovered, setIsHovered] = useState(false)
  const [isSubmitting, setIsSubmitting] = useState(false)

  function formatLikeText(count: number): string {
    if (count === 0) return 'Curtir'
    return count === 1 ? '1 curtida' : `${count} curtidas`
  }

  async function toggleLike() {
    if (isSubmitting) return

    setIsSubmitting(true)

    const newLikeState = !isLiked

    setIsLiked(newLikeState)
    setLikeCount((prevCount) =>
      newLikeState ? prevCount + 1 : Math.max(0, prevCount - 1),
    )

    if (!newLikeState) {
      setIsHovered(false)
    }

    try {
      const response = await fetch(
        `${process.env.NEXT_PUBLIC_NEXTAUTH_URL}/api/comments/${commentId}/likes`,
        {
          method: 'PUT',
        },
      ).then((res) => res.json())

      if (response.isLike !== newLikeState) {
        setIsLiked(response.isLike)
        setLikeCount((prevCount) =>
          response.isLike ? prevCount + 1 : Math.max(0, prevCount - 1),
        )
      }

      if (response.error) {
        setIsLiked(!newLikeState)
        setLikeCount((prevCount) =>
          !newLikeState ? prevCount + 1 : Math.max(0, prevCount - 1),
        )

        toaster.create({
          description: response.error,
          type: 'error',
          duration: 1500,
        })
      }
    } catch (error) {
      setIsLiked(!newLikeState)
      setLikeCount((prevCount) =>
        !newLikeState ? prevCount + 1 : Math.max(0, prevCount - 1),
      )

      toaster.create({
        description: 'Erro ao processar sua curtida',
        type: 'error',
        duration: 1500,
      })
    } finally {
      setIsSubmitting(false)
    }
  }

  return (
    <button
      {...props}
      onClick={toggleLike}
      onMouseEnter={() => setIsHovered(true)}
      onMouseLeave={() => {
        if (!isLiked) {
          setIsHovered(false)
        }
      }}
      disabled={isSubmitting}
      aria-label={isLiked ? 'Remover curtida' : 'Curtir tÃ³pico'}
      aria-pressed={isLiked}
      className="flex items-center gap-2 transition-colors focus:outline-none"
    >
      <HeartIcon isActive={isLiked} isHovered={isHovered} />
      <span className="text-xs">{formatLikeText(likeCount)}</span>
    </button>
  )
}
