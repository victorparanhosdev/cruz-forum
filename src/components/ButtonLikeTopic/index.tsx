'use client'
import { Heart } from '@phosphor-icons/react/dist/ssr'
import { ButtonHTMLAttributes, useState } from 'react'
import { toaster } from '../ui/toaster'
import { motion } from 'framer-motion'

interface ButtonLikeTopicProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  likes: number
  topicSlug: number
  isLike: boolean
}

function IconHeartComponent({ isLike }: { isLike: boolean }) {
  const [isHovered, setIsHovered] = useState(false)

  return (
    <motion.div
      initial={false}
      animate={{ scale: isLike || isHovered ? 1.1 : 1 }}
      transition={{ type: 'spring', stiffness: 300, damping: 20 }}
      onMouseEnter={() => !isLike && setIsHovered(true)}
      onMouseLeave={() => !isLike && setIsHovered(false)}
    >
      <Heart
        className="text-white"
        size={20}
        aria-hidden="true"
        weight={isLike ? 'fill' : isHovered ? 'fill' : 'regular'}
      />
    </motion.div>
  )
}

export const ButtonLikeTopic = ({
  likes,
  isLike,
  topicSlug,
  ...props
}: ButtonLikeTopicProps) => {
  const [isLikeTopic, setIsLikeTopic] = useState(isLike)
  const [counterLikes, setCounterLikes] = useState<number>(likes)

  async function handleLikeTopic() {
    try {
      const response = await fetch(
        `${process.env.NEXT_PUBLIC_NEXTAUTH_URL}/api/topics/${topicSlug}/likes`,
        {
          method: 'PUT',
        },
      ).then((res) => res.json())

      if (response.isLike) {
        setCounterLikes((prev) => prev + 1)
        setIsLikeTopic(true)
        return
      }

      if (response.error) {
        toaster.create({
          description: response.error,
          type: 'error',
          duration: 1500,
        })
        return
      }

      setCounterLikes((prev) => prev - 1)
      setIsLikeTopic(false)
    } catch (error) {
      toaster.create({
        description: 'Error ao curtir topico',
        type: 'error',
        duration: 1500,
      })
    }
  }

  return (
    <button
      {...props}
      onClick={handleLikeTopic}
      aria-label="BotÃ£o de curtir"
      className="flex items-center gap-2"
    >
      <IconHeartComponent isLike={isLikeTopic} /> {counterLikes} curtidas
    </button>
  )
}
