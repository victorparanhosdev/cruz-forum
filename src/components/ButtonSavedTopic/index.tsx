'use client'
import { BookmarkSimple } from '@phosphor-icons/react'
import { HtmlHTMLAttributes, useState } from 'react'
import { toaster } from '../ui/toaster'
import { usePathname, useRouter } from 'next/navigation'
import { motion } from 'framer-motion'
interface ButtonSavedTopic extends HtmlHTMLAttributes<HTMLButtonElement> {
  topicSlug: number
  isSavedTopic: boolean
}

function IconBookMarkSimple({ isSaved }: { isSaved: boolean }) {
  const [isHovered, setIsHovered] = useState(false)

  return (
    <motion.div
      initial={false}
      animate={{ scale: isSaved || isHovered ? 1.1 : 1 }}
      transition={{ type: 'spring', stiffness: 300, damping: 20 }}
      onMouseEnter={() => !isSaved && setIsHovered(true)}
      onMouseLeave={() => !isSaved && setIsHovered(false)}
    >
      <BookmarkSimple
        className="text-white"
        size={36}
        aria-hidden="true"
        weight={isSaved ? 'fill' : isHovered ? 'fill' : 'regular'}
      />
    </motion.div>
  )
}
export const ButtonSavedTopic = ({
  topicSlug,
  isSavedTopic,
  ...props
}: ButtonSavedTopic) => {
  const [isSaved, setSaved] = useState(isSavedTopic)
  const pathname = usePathname()
  const router = useRouter()

  async function handleSavedTopic() {
    const response = await fetch(
      `${process.env.NEXT_PUBLIC_NEXTAUTH_URL}/api/topics/${topicSlug}/saved`,
      {
        method: 'PUT',
      },
    ).then((res) => res.json())

    if (response.isSaved) {
      toaster.create({
        description: 'Tópico salvo com sucesso!',
        type: 'info',
        duration: 1500,
      })

      return setSaved(true)
    }

    if (!response.iSaved && pathname === '/salvos') {
      router.refresh()
    }

    toaster.create({
      description: "Tópico removido de 'Salvos'",
      type: 'error',
      duration: 1500,
    })

    setSaved(false)
  }

  return (
    <button {...props} onClick={handleSavedTopic}>
      <IconBookMarkSimple isSaved={isSaved} />
    </button>
  )
}
