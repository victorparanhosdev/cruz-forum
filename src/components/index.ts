export { AlertDialog } from './AlertDialog'
export { Button } from './Button'
export { Card } from './Card'
export { Comentarios } from './Comentarios'
export { Input, InputMsgErro } from './InputText'
export { Label } from './Label'
export { Navigation } from './Navigation'
export { Menu } from './Navigation/menu'
export { PopoverPerfil } from './PopoverPerfil'
export { Topic } from './Topic'
export { CloseButton, type CloseButtonProps } from './ui/close-button'
export { TextArea, TextAreaMsgErro } from './TextArea'
export { FilterPopover } from './FilterPopover'
export { TopicDialog } from './TopicDialog'
export { AlertDialogDeleteComment } from './AlertDialogDeleteComment'
export { ButtonSavedTopic } from './ButtonSavedTopic'
export { ButtonLikeTopic } from './ButtonLikeTopic'
export * as Popover from './ui/popover'
