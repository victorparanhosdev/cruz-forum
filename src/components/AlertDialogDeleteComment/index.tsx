'use client'
import {
  DialogActionTrigger,
  DialogBody,
  DialogCloseTrigger,
  DialogContent,
  DialogFooter,
  DialogHeader,
  DialogRoot,
  DialogTitle,
  DialogTrigger,
} from '@/components/ui/dialog'
import { ComponentProps, ReactNode, useState } from 'react'
import { Button } from '@/components'
import { toaster } from '@/components/ui/toaster'
import { CircleNotch } from '@phosphor-icons/react'
import { useRouter } from 'next/navigation'

interface AlertDialogProps extends ComponentProps<typeof DialogTrigger> {
  children: ReactNode
  commentId: string
}

export const AlertDialogDeleteComment = ({
  commentId,
  children,
  ...props
}: AlertDialogProps) => {
  const [isOpen, setIsOpen] = useState(false)
  const [isLoading, setIsLoading] = useState(false)
  const router = useRouter()

  async function handleDeleteComment() {
    setIsLoading(true)
    try {
      const response = await fetch(
        `${process.env.NEXT_PUBLIC_NEXTAUTH_URL}/api/comments/${commentId}/delete`,
        { method: 'DELETE' },
      )

      if (!response.ok) {
        const errorData = await response.json()

        if (errorData.error) {
          setIsLoading(false)
          setIsOpen(false)

          return toaster.error({
            title: 'Error',
            description: errorData.error,
            duration: 2000,
            type: 'error',
          })
        }

        return errorData
      }

      const data = await response.json()

      router.refresh()
      setIsLoading(false)
      setIsOpen(false)

      toaster.success({
        title: 'Deletado com sucesso!',
        description: data?.message || 'O comentario foi removido com sucesso.',
        duration: 1500,
      })

      return data
    } catch (error) {
      console.log('Deu errado:', error)
      return { error: 'Erro desconhecido ao excluir o comentario.' }
    }
  }

  const handleDialogClose = ({ open }: { open: boolean }) => {
    setIsOpen(open)
  }

  return (
    <DialogRoot
      open={isOpen}
      onOpenChange={handleDialogClose}
      placement="center"
      role="alertdialog"
    >
      <DialogTrigger {...props} className="flex cursor-pointer items-center">
        {children}
      </DialogTrigger>
      <DialogContent
        className="mx-6 rounded-lg bg-gray-950 md:mx-auto"
        padding={6}
      >
        <DialogHeader padding={0} marginBottom={4}>
          <DialogTitle className="text-lg font-medium sm:text-xl md:text-2xl">
            Excluir Coment√°rio
          </DialogTitle>
        </DialogHeader>

        <DialogBody padding={0}>
          <h1 className="text-xs text-zinc-300 sm:text-sm md:text-base">
            Tem certeza de que deseja excluir o comentario?
          </h1>
        </DialogBody>

        <DialogFooter padding={0} marginTop={6}>
          <DialogActionTrigger asChild>
            <Button state="outline-negative">Cancelar</Button>
          </DialogActionTrigger>

          {isLoading ? (
            <Button className="min-h-11 min-w-[101.16px]">
              <CircleNotch
                className="h-full w-full animate-spin text-white"
                size={20}
              />
            </Button>
          ) : (
            <Button type="button" onClick={handleDeleteComment}>
              Continue
            </Button>
          )}
        </DialogFooter>

        <DialogCloseTrigger />
      </DialogContent>
    </DialogRoot>
  )
}
