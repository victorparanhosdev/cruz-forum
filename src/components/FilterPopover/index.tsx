import { Popover } from '@/components'
import { ReactNode } from 'react'
import {
  CaretUpDown,
  ChatTeardropText,
  Heart,
  StarFour,
  TrendUp,
} from '@phosphor-icons/react/dist/ssr'

export const FilterPopover = ({ children }: { children: ReactNode }) => {
  return (
    <Popover.PopoverRoot positioning={{ sameWidth: true }}>
      <Popover.PopoverTrigger asChild>{children}</Popover.PopoverTrigger>
      <Popover.PopoverContent
        className="rounded-lg border border-stone-700 bg-zinc-950"
        overflow="hidden"
        width="fit-content"
      >
        <Popover.PopoverBody padding="initial">
          <ul className="bg-zinc-950 text-sm">
            <li className="flex w-full cursor-pointer items-center gap-2.5 whitespace-nowrap px-4 py-2.5 transition hover:bg-hover-btn-menu_card ">
              <StarFour size={20} />
              Recentes
            </li>
            <li className="flex w-full cursor-pointer items-center gap-2.5 whitespace-nowrap px-4 py-2.5 transition hover:bg-hover-btn-menu_card ">
              <TrendUp size={20} />
              Mais Relevantes
            </li>
            <li className="flex w-full cursor-pointer items-center gap-2.5 whitespace-nowrap px-4 py-2.5 transition hover:bg-hover-btn-menu_card ">
              <ChatTeardropText size={20} />
              Mais comentados
            </li>
            <li className="flex w-full cursor-pointer items-center gap-2.5 whitespace-nowrap px-4 py-2.5 transition hover:bg-hover-btn-menu_card ">
              <Heart size={20} />
              Mais curtidos
            </li>
            <li className="flex w-full cursor-pointer items-center gap-2.5 whitespace-nowrap px-4 py-2.5 transition hover:bg-hover-btn-menu_card ">
              <CaretUpDown size={20} />
              Ordenar A-Z
            </li>
          </ul>
        </Popover.PopoverBody>
      </Popover.PopoverContent>
    </Popover.PopoverRoot>
  )
}
