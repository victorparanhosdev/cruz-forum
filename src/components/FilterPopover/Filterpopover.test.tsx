import { render, screen, fireEvent, waitFor } from '@testing-library/react'
import { Button, FilterPopover } from '@/components'
import { ChakraUIProvider } from '../../providers/chakra-ui'

global.ResizeObserver = class {
  observe() {}
  unobserve() {}
  disconnect() {}
}
const FilterPopoverProvider = (
  <ChakraUIProvider>
    <FilterPopover>
      <Button data-testid="trigger-button">Abrir Filtro</Button>
    </FilterPopover>
  </ChakraUIProvider>
)

describe('<FilterPopover />', () => {
  it('deve abrir o popover ao clicar no trigger e fechar ao clicar novamente', async () => {
    render(FilterPopoverProvider)
    const buttonElement = screen.getByTestId('trigger-button')

    fireEvent.click(buttonElement)

    await waitFor(() => {
      expect(buttonElement).toHaveAttribute('data-state', 'open')
    })
    fireEvent.click(document.body)

    await waitFor(() => {
      expect(buttonElement).not.toHaveAttribute('data-state', 'closed')
    })
  })

  it('deve corresponder ao snapshot', () => {
    const { asFragment } = render(FilterPopoverProvider)

    expect(asFragment()).toMatchSnapshot()
  })
})
