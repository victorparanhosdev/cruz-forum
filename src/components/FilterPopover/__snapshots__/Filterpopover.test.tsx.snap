// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilterPopover /> deve corresponder ao snapshot 1`] = `
<DocumentFragment>
  <button
    aria-controls="popover::r1::content"
    aria-expanded="false"
    aria-haspopup="dialog"
    class="flex items-center justify-center gap-2 rounded-lg text-base font-medium transition bg-green-950 px-4 py-2.5 enabled:hover:bg-green-800 disabled:bg-gray-400 disabled:text-gray-700 chakra-popover__trigger css-0"
    data-part="trigger"
    data-scope="popover"
    data-state="closed"
    data-testid="trigger-button"
    dir="ltr"
    id="popover::r1::trigger"
    type="button"
  >
    <span
      class="whitespace-nowrap text-base"
    >
      Abrir Filtro
    </span>
  </button>
</DocumentFragment>
`;
